api/maven-api-core/src/main/java/org/apache/maven/api/services/Source.java

```diff
@@ -20,17 +20,54 @@
 
 import java.io.IOException;
 import java.io.InputStream;
+import java.nio.file.Path;
 
+import org.apache.maven.api.Session;
 import org.apache.maven.api.annotations.Experimental;
+import org.apache.maven.api.annotations.Nonnull;
+import org.apache.maven.api.annotations.Nullable;
 
 /**
- * The source for a project's XML model.
+ * Provides access to the contents of a source independently of the
+ * backing store (e.g. file system, database, memory).
+ * <p>
+ * This is mainly used to parse files into objects such as
+ * {@link org.apache.maven.api.Project},
+ * {@link org.apache.maven.api.model.Model},
+ * {@link org.apache.maven.api.settings.Settings}, or
+ * {@link org.apache.maven.api.toolchain.PersistedToolchains}.
  *
  * @since 4.0.0
+ * @see org.apache.maven.api.services.ProjectBuilder#build(Session, Source)
+ * @see org.apache.maven.api.services.SettingsBuilder#build(Session, Source, Source, Source)
+ * @see org.apache.maven.api.services.ToolchainsBuilder#build(Session, Source, Source)
  */
 @Experimental
 public interface Source {
+
+    /**
+     * Provides access the file to be parsed, if this source is backed by a file.
+     *
+     * @return The underlying {@code Path}, or {@code null} if this source is not backed by a file.
+     */
+    @Nullable
+    Path getPath();
+
+    /**
+     * Gets a byte stream to the source contents. Closing the returned stream is the responsibility of the caller.
+     *
+     * @return A byte stream to the source contents, never {@code null}.
+     * @throws IOException in case of IO issue
+     */
+    @Nonnull
```

Comment:
i would call this method `openStream()` to make it more clear that this always creates a new stream and is not a simple getter.